File: 1717-maximum-score-from-removing-substrings/1717-maximum-score-from-removing-substrings.cpp
class Solution {
private:
    int removeSubstr(string& s, string target, int ptPerRemoval){
        int writeIdx = 0, total = 0;
        
        // Iterate through the string
        for (int readIdx = 0; readIdx < s.size(); ++readIdx){
            // Add the current character
            s[writeIdx++] = s[readIdx];
            
            // Check if we've written at least two characters and
            // they match the target substring
            if (writeIdx > 1 && s[writeIdx - 2] == target[0] &&\
               s[writeIdx - 1] == target[1]){
                writeIdx -= 2; // Move write index back to remove the match
                total += ptPerRemoval;
            }
        }
        
        // Trim the string to remove any leftover characters
        s.erase(s.begin() + writeIdx, s.end());
        
        return total;
    }
        
public:
    int maximumGain(string s, int x, int y) {
        int ans = 0;
        
        if (x > y){
            // Remove "ab" first (higher points), then "ba" 
            ans += removeSubstr(s, "ab", x);
            ans += removeSubstr(s, "ba", y);
        }
        else{
            // Remove "ba" first (higher or equal points), then "ab"
            ans += removeSubstr(s, "ba", y);
            ans += removeSubstr(s, "ab", x);
        }
        
        return ans;
    }
};

File: 1717-maximum-score-from-removing-substrings/1717-maximum-score-from-removing-substrings.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        ans, s = 0, list(s)
        
        if x > y:
            # Remove "ab" first (higher points), then "ba"
            ans += self.remove_substr(s, "ab", x)
            ans += self.remove_substr(s, "ba", y)
        else:
            # Remove "ba" first (higher, or equal points), then "ab"
            ans += self.remove_substr(s, "ba", y)
            ans += self.remove_substr(s, "ab", x)
        
        return ans
    
    def remove_substr(self, input: str, target: str, points_per_removal: int) -> str:
        total, write_idx = 0, 0
        
        # Iterate through the string
        for read_idx in range(len(input)):
            # Add the current character"
            input[write_idx] = input[read_idx]
            write_idx += 1
            
            # Check if we've written at least 2 characters and
            # they match the target substring
            if (write_idx > 1 and input[write_idx - 2] == target[0] and\
                input[write_idx - 1] == target[1]):
                write_idx -= 2
                total += points_per_removal
                
        # Trim the list to remove any leftover characters
        del input[write_idx:]
        
        return total